src/puppetlabs/trapperkeeper/internal.clj

```diff
@@ -193,15 +201,37 @@
       (log/errorf t "Error during service %s!!!" lifecycle-fn-name)
       (throw t))))
 
-(def ^:private sighup-lock (Object.))
+(schema/defn ^:always-validate initialize-lifecycle-worker :- (schema/protocol async-prot/Channel)
+  "Initializes a 'worker' which will listen for lifecycle-related tasks and perform
+  them on a background thread, to ensure that we aren't executing multiple lifecycle
+  tasks simultaneously."
+  [lifecycle-chan :- (schema/protocol async-prot/Channel)
+   shutdown-reason-promise :- IDeref]
+  (async/go (try
+              (loop []
+                (let [task (async/<! lifecycle-chan)]
+                  (if (= task :shutdown)
+                    (do
+                      (log/debug "Received shutdown command, lifecycle worker exiting.")
+                      :done)
+                    (do
+                      (log/debug "Lifecycle worker executing lifecycle task.")
+                      (task)
+                      (log/debug "Lifecycle worker completed lifecycle task; awaiting next task.")
+                      (recur)))))
+              (catch Exception e
+                (deliver shutdown-reason-promise
+                         {:cause :service-error
+                          :error e})))))
 
 (defn restart-tk-apps
   "Call restart on all tk apps."
   [apps]
   (log/debug "SIGHUP handler restarting TK apps.")
-  (locking sighup-lock
-    (doseq [app apps]
-      (a/restart app))))
+  (doseq [app apps]
+    (let [{:keys [lifecycle-channel]} @(a/app-context app)
+          restart-fn #(a/restart app)]
+      (async/>!! lifecycle-channel restart-fn))))
```

Comment:
Should all of the apps that need to be restarted be restarted within one "task" handled on the lifecycle channel - as opposed to one "task" per app as it would appear to be done here.  Not sure if there would be issues at some point with other tasks potentially being interleaved between individual app restarts that could cause a problem - even if not now, then maybe for some future lifecycle task that we might introduce?